### ISSUE TITLE
apply_async hangs sporadically with high CPU usage when connecting to RabbitMQ

### ISSUE THREAD

**Issue by tyarimi** [celery/celery#4167]
We are using celery 4.1.0, kombu 4.1.0, vine 1.1.4, amqp 2.2.1, billiard 3.5.0.3.

We have some long running processes that enqueue tasks on RabbitMQ. After running for a few hours, some of them start hanging forever with high CPU usage. Here is a stacktrace of one of them (last call first):

```
File "/usr/local/lib/python2.7/dist-packages/vine/five.py", line 203 in array
  File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 252 in __init__
  File "/usr/local/lib/python2.7/dist-packages/kombu/transport/pyamqp.py", line 128 in establish_connection
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 757 in _establish_connection
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 802 in connection
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 261 in connect
  File "/usr/local/lib/python2.7/dist-packages/kombu/utils/functional.py", line 333 in retry_over_time
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 405 in ensure_connection
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 515 in _ensured
  File "/usr/local/lib/python2.7/dist-packages/kombu/common.py", line 147 in _imaybe_declare
  File "/usr/local/lib/python2.7/dist-packages/kombu/common.py", line 128 in maybe_declare
  File "/usr/local/lib/python2.7/dist-packages/celery/backends/rpc.py", line 168 in on_task_call
  File "/usr/local/lib/python2.7/dist-packages/celery/app/base.py", line 736 in send_task
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 536 in apply_async
...
```

Seems like it gets stuck when trying to connect to Rabbit, maybe in an infinite loop inside the `ensure_connection` function.

Here is our celery config:

```python
CELERY_BACKEND = 'rpc'
BROKER_HEARTBEAT = 60
CELERYD_HIJACK_ROOT_LOGGER = False
CELERY_RESULT_BACKEND = CELERY_BACKEND
CELERY_ACCEPT_CONTENT = {'pickle'}
CELERY_TRACK_STARTED = True
CELERY_TIMEZONE = 'UTC'
CELERYD_MAX_TASKS_PER_CHILD = 1000
CELERY_SEND_EVENTS = False
CELERY_EVENT_QUEUE_EXPIRES = 60
CELERYD_TASK_SOFT_TIME_LIMIT = 12 * 60 * 60  # 12 hours
CELERY_RESULT_SERIALIZER = CELERY_TASK_SERIALIZER = 'pickle'
CELERYD_CONCURRENCY = 5
CELERYD_PREFETCH_MULTIPLIER = 1
```

Any ideas?

**Comment by matteius** [comment]
I just found this ticket while researching a similar issue that started happening after a software release.   We didn't change celery versions but we did change the results backend from amqp to rpc which is also the setting you call out.   Our versions are as follows but plans to upgrade to the versions you report are in our backlog:
celery==4.0.0
kombu==4.0.0
amqp==2.1.1
billiard==3.5.0.2

**Comment by tyarimi** [comment]
@matteius do you have a stacktrace?

**Comment by tyarimi** [comment]
Seems like the source of the problem is in the `[on_task_call](https://github.com/celery/celery/blob/master/celery/backends/rpc.py#L168)` function in rpc.py. It calls `maybe_declare` with `retry=True` but won't pass a "retry policy", specifically the "max_retries" parameter. This means the _ensure function down the stack can go into an infinite loop.

@ask - does this make sense? Should we add max_retries to this function?

**Comment by matteius** [comment]
I do not have a stack trace because it doesn't actually crash the application, but I have verified the issue is with the rpc remote back and not the deprecated but still existant amqp backend.  I am able to relay these facts to support my claim:
1.) A software release changed the setting CELERY_RESULT_BACKEND (result_backend) in 8 projects from 'amqp' to 'rpc' to end a deprecation warning (rpc class uses amqp behind the scenes).
2.) After said release, about 2.5 hours or so, one of the services started a step function that would eventually overtake the CPU and bring down the box.
3.) After intense analysis we determined that the step function was individual uwsgi processes eventually hanging and consuming a step up of CPU activity.  This was correlated to a massive increase in the number of connections to the Rabbit MQ cluster which was alarming to some people.
4.) I reverted the change to the result backend for the affected service and CPU behavior returned to normal.

For what its worth, restarting uwsgi generally brought back down the CPU activity and resumed behavior but the timing and step function was consistent on the rpc backend.  It never actually caused the application to crash so there was never a stack trace.  It was also hard to determine it was the change to the backend because it was not clear that any other service was affected in the same way.  

We generally aren't using results backend anyway so I after re-reading the settings docs I realize now that the default is not to have at results backend and we may try removing that setting from projects (or setting it to None) that do not check for results of tasks.

I do not have a stack trace because it doesn't actually crash the application, but I have verified the issue is with the rpc remote back and not the deprecated but still existant amqp backend.  I am able to relay these facts to support my claim:
1.) A software release changed the setting CELERY_RESULT_BACKEND (result_backend) in 8 projects from 'amqp' to 'rpc' to end a deprecation warning (rpc class uses amqp behind the scenes).
2.) After said release, about 2.5 hours or so, one of the services started a step function that would eventually overtake the CPU and bring down the box.
3.) After intense analysis we determined that the step function was individual uwsgi processes eventually hanging and consuming a step up of CPU activity.  This was correlated to a massive increase in the number of connections to the Rabbit MQ cluster which was alarming to some people.
4.) I reverted the change to the result backend for the affected service and CPU behavior returned to normal.

For what its worth, restarting uwsgi generally brought back down the CPU activity and resumed behavior but the timing and step function was consistent on the rpc backend.  It never actually caused the application to crash so there was never a stack trace.  It was also hard to determine it was the change to the backend because it was not clear that any other service was affected in the same way.  

We generally aren't using results backend anyway so I after re-reading the settings docs I realize now that the default is not to have at results backend and we may try removing that setting from projects (or setting it to None) that do not check for results of tasks.

I have a theory as to the timing of the step function but I think that was caused by another network issue and so shouldn't be considered in this fix.  I definitely think you found the bug with the retry policy now that I checked that link to the src.

**Comment by thedrow** [comment]
@tyarimi I believe your analysis is correct. Care to issue a pull request?

**Comment by tyarimi** [comment]
Actually I found the real bug in Kombu and issued [a PR there](https://github.com/celery/kombu/pull/779). No response yet.

**Comment by tyarimi** [comment]
We have been running with the fix in a large production environment for over a year now, and it works perfectly.
It wasn't merged since it doesn't have tests, which I did not implement as I'm not sure how, using the current test environment.

**Comment by adib** [comment]
I suppose for 4.2 the workaround is to use the same AMQP URL as `backend` as well as `broker` ?

**Comment by matteius** [comment]
I could _imaybe_declare try to write a unit test using mock on a fork on your kombu PR once I figure out what that code is actually for.

**Comment by tyarimi** [comment]
@matteius that will be great. You can find more details in the [PR conversation](https://github.com/celery/kombu/pull/779#discussion_r132499616).

**Comment by tyarimi** [comment]
@matteius regarding the settings - yes we do use the old notation which is still supported, and it works. We run millions of tasks every day without issues. Maybe we have some old leftover settings which are ignored by Celery, but it doesn't affect anything.

I was wondering about the `_i` prefix myself, maybe it means `internal`.

I'm absolutely positive that this is a logical bug in Kombu, and that it's not caused due to incorrect usage. From my [detailed explanation](https://github.com/celery/kombu/pull/779#discussion_r134961611) we see that the problem starts when the `rpc` backend is calling `maybe_declare` with the default `channel=None` (which is surely supported, as this is the default value of that function). From there the entire logic is inside Kombu.

The reason this bug is very rare and hard to reproduce is because you need the channel to get broken in a very specific point. For us this was happening only once in a few millions of tasks. I reproduced it locally by changing the code to drop the connection just once at a specific point (I don't have this code anymore since it was a long time ago), then even when the connection is back up, Kombu is already inside an infinite loop which, locally, cannot be exited.

**Comment by jackejiang** [comment]
We encountered the similar issue when the heartbeat of rabbitmq connection timeouts and the connection is lost.  To reproduce the issue, we could manually run the python console by following steps:

1, Run below lines:
>>> from celery import Celery
>>> app = Celery('task', broker='amqp://test:test@x.x.x.x/test', backend='rpc://')
>>> r = app.send_task('add', args=[1,2], queue='test_q')

2, From rabbitmq management UI,  force close the connection established

3, Run an extra line, and it hangs:
>>> r = app.send_task('add', args=[1,2], queue='test_q')

From rabbitmq log, we could see a lot of connection logs, and from python console, after ctrl+c, get  below backtrace:
File "/usr/lib/python2.7/site-packages/celery/app/base.py", line 744, in send_task
    self.backend.on_task_call(P, task_id)
  File "/usr/lib/python2.7/site-packages/celery/backends/rpc.py", line 168, in on_task_call
    maybe_declare(self.binding(producer.channel), retry=True)
  File "/usr/lib/python2.7/site-packages/kombu/common.py", line 128, in maybe_declare
    channel, orig, **retry_policy)
  File "/usr/lib/python2.7/site-packages/kombu/common.py", line 147, in _imaybe_declare
    entity, declared, ident, channel, orig)
  File "/usr/lib/python2.7/site-packages/kombu/connection.py", line 515, in _ensured
    reraise_as_library_errors=False,
  File "/usr/lib/python2.7/site-packages/kombu/connection.py", line 405, in ensure_connection
    callback)
  File "/usr/lib/python2.7/site-packages/kombu/utils/functional.py", line 332, in retry_over_time
    return fun(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/kombu/connection.py", line 261, in connect
    return self.connection
  File "/usr/lib/python2.7/site-packages/kombu/connection.py", line 802, in connection
    self._connection = self._establish_connection()
  File "/usr/lib/python2.7/site-packages/kombu/connection.py", line 757, in _establish_connection
    conn = self.transport.establish_connection()
  File "/usr/lib/python2.7/site-packages/kombu/transport/pyamqp.py", line 128, in establish_connection
    conn = self.Connection(**opts)
  File "/usr/lib/python2.7/site-packages/amqp/connection.py", line 258, in __init__
    self._avail_channel_ids = array('H', range(self.channel_max, 0, -1))
  File "/usr/lib/python2.7/site-packages/vine/five.py", line 203, in array
    return _array(typecode, *args, **kwargs)

We tested in 4.2.0 and 4.2.1, and got the same result

**Comment by thedrow** [comment]
The fix is now merged and scheduled for release on 4.3.
